Rewrite framed replay input handler for robustness

ekrctb opened the pull request:
### Motivation

- The previous code was full of nullable types and it was cumbersome to use (by the derived classes).
- There was no explicit invariant and it was difficult to reason about.
- Fixes #6150. Fixes #10811.

### Methods

I tried to make less breaking changes as possible, but there are some changes.
The getters of the frames now throw when the replay is empty, instead of returning `null`.
The values of `CurrentFrame` and `NextFrame` are changed. Namely, the ordering of `NextFrame.Time` and `CurrentFrame.Time` was dependent on the current direction. Now, it should always satisfy `CurrentFrame.Time <= CurrentTime <= NextFrame.Time` except at the start/end.

### Result

- The issue #6150 can be visually confirmed by repeatedly rewinding a catch replay and noticing the catcher goes out-of-bounds in master but that is fixed with this PR.
- Luckily, the change of `CurrentFrame` and `NextFrame` doesn't break existing deriving classes if the template code pattern usage of interpolation is used.
  - This is due to the symmetry: `interpolate(time, start, end) = interpolate(time, end, start)`.
- The deriving class code can be simplified due to the elimination of nullable types. This is <https://github.com/ppy/osu/commit/bfb7805838be2e777b27919a127040597b7092c0> but not included in this PR because a breaking change is required to simplify more and I would like to have a separate PR.
  - Edit: this change is partially reverted, as it was breaking a certain case. It will be included in the next PR.
- Return value of `SetFrameFromTime` (whether to return null) is also changed, but it only affected the testing code, and also it follows the documentation (returning `null` when the frame is not valid) more correctly.

### Discussion

I wanted to make `ReplayInputHandler` is not called when the replay is empty, using `IsActive` property of `InputHandler`. However, `IsActive` is unimplemented in the osu-framework (dead code) so I had to modify `RulesetInputManager`. Probably it should be supported in `osu-framework`, or `IsActive` shouldn't be used.
`StreamingFramedReplayInputHandlerTest` is removed, as it was duplicate. I added a test to `FramedReplayInputHandlerTest` as a replacement.

ekrctb committed:
Throw when getting a frame of an empty replay

ekrctb committed:
Rewrite FramedReplayInputHandler for robustness

This commit changes the semantics of `CurrentFrame` and `NextFrame` of the class.
The ordering of `NextFrame.Time` and `CurrentFrame.Time` was dependent on the current direction.
Now, it should always satisfy `CurrentFrame.Time <= CurrentTime <= NextFrame.Time` except at the start/end.
This change, however, doesn't break existing deriving classes if the template code pattern usage of interpolation is used.
The deriving class code can be simplified due to the elimination of nullable types. I didn't include those changes in this commit.

I removed `StreamingFramedReplayInputHandlerTest` for now, as it is almost-duplicate of `FramedReplayInputHandlerTest`. I'll include more tests in later commits.

This commit fixes #6150.

ekrctb committed:
Add more FramedReplayInputHandler tests

peppy commented:
Curious if this happens to resolve https://github.com/ppy/osu/issues/10811 as well.

ekrctb committed:
Remove `#nullable true` for now to suppress inspector

peppy commented on osu.Game/Rulesets/UI/RulesetInputManager.cs:
```diff
@@ -100,6 +102,14 @@ public ReplayInputHandler ReplayInputHandler
+        protected override List<IInput> GetPendingInputs()
+        {
+            if (replayInputHandler != null && !replayInputHandler.IsActive)
+                return new List<IInput>();
```
The type of this method is unfortunate in that we can't use a static representation of the empty list. To avoid allocation of the `new List<>` each call, a better option may be to call the base method and then perform a `.Clear()` on the list.

peppy commented on osu.Game/Rulesets/Replays/FramedReplayInputHandler.cs:
```diff
@@ -17,65 +16,80 @@ namespace osu.Game.Rulesets.Replays
+        /// <summary>
+        /// Whether we are waiting for new frames to be received.
+        /// </summary>
+        public bool WaitingNextFrame => !replay.HasReceivedAllFrames && currentFrameIndex == Frames.Count - 1;
```
```suggestion
        public bool WaitingForFrame => !replay.HasReceivedAllFrames && currentFrameIndex == Frames.Count - 1;
```

or `WaitingForNextFrame`.

peppy commented on osu.Game/Rulesets/Replays/FramedReplayInputHandler.cs:
```diff
@@ -17,65 +16,80 @@ namespace osu.Game.Rulesets.Replays
 
+        /// <summary>
+        /// The next frame of the replay.
+        /// The start time is always greater or equals to the start time of <see cref="CurrentFrame"/> regardless of the seeking direction.
```
```suggestion
        /// The start time is always greater or equal to the start time of <see cref="CurrentFrame"/> regardless of the seeking direction.
```

peppy commented on osu.Game/Rulesets/Replays/FramedReplayInputHandler.cs:
```diff
@@ -17,65 +16,80 @@ namespace osu.Game.Rulesets.Replays
+        protected FramedReplayInputHandler(Replay replay)
+        {
+            // This replay frame ordering should be enforced on the Replay type
+            replay.Frames.Sort((x, y) => x.Time.CompareTo(y.Time));
```
Is the intention to remove this eventually (in a follow up PR)? Probably best to prefix the comment with `TODO:`.

Also as this is done at construction time, it won't be applied to streaming frames so that's something to consider.

peppy reviewed (changes requested):
Just an initial pass. Will do some more reading on the actual logic changes and your questions about the `InputManager.IsActive` usage in a second pass.

ekrctb committed:
Avoid empty list allocation

ekrctb committed:
Rename property

ekrctb committed:
Add TODO comment

ekrctb commented:
The `TestSceneGameplayRewinding` failure is due to the change of `CurrentFrame` (while rewinding).
`KeyCounter` has `IsCounting` property. This is set to false when the gameplay is in a break and set to true when no longer in the break, but the change is delaying one frame. The ordering of the process is as follows (when rewinding from the end of the map):

1. (New frame)
2. `BreakTracker.Update`: `IsBreakTime = true` because `HasCompleted == true`.
3. `RulesetInputManager` process input actions. It triggers `KeyCounter.OnReleased`, but it is not counted as `IsCounting == false`, which is bound to `IsBreakTime`.
4. `DHO.Update` triggers `RevertResult`, its event listener set `HasCompleted = false`.
5. (New frame)

Previously, the future frame's input actions were used while rewinding, which is semantically wrong (this issue is different from #6150), but the key counter worked because it effectively delayed the input action change to compensate for the one-frame delay above.

Fixing it by changing drawable child ordering seems difficult and fragile.
I think a proper solution is to make `KeyCounter` record past key presses in an explicit list (potentially using `Replay`, but not sure it makes sense) so state rewinding can be done robustly. It would be a separate PR.

ekrctb committed:
Merge branch 'master' into refactor-framed-replay-input-hander

ekrctb committed:
Delay map completion one frame after the last judgment

This is a workaround of a timing issue.
KeyCounter is disabled while break time (`HasCompleted == true`).
When the last keypress is exactly at the same time the map ends, the last frame was considered in a break time while forward play but considered not in a break time while rewinding. This inconsistency made the last keypress not decremented in the key counter when a replay is rewound.
The situation regularly happens in osu!standard because the map ends right after the player hits the last hit circle. It was caught by `TestSceneGameplayRewinding`.

This commit makes the update of the map completion delayed one frame. The problematic keypress frame is now processed strictly before the map completion, and the map completion status is correctly rewound before the keypress frame.

ekrctb commented:
Hopefully I managed to fix the test failure with a little code change. See the commit message for the elaboration.

ekrctb committed:
Partially revert the changes of `CurrentFrame` and `NextFrame` for compatibility

Making those always non-null is postponed as when a replay's frame contains keypress the behavior is changed.
Previously, the key is pressed at the time of the first frame. But using non-null frames means the key is pressed at negative infinity.
However, I think the new way of always using non-null frames makes the client code so I plan to bundle the change to more breaking changes.

peppy commented on osu.Game/Rulesets/Scoring/JudgementProcessor.cs:
```diff
@@ -134,6 +133,10 @@ void simulate(HitObject obj)
+        protected override void Update()
+        {
+            base.Update();
+            hasCompleted.Value = JudgedHits == MaxHits && (JudgedHits == 0 || lastAppliedResult.TimeAbsolute < Clock.CurrentTime);
```
Personally I'm okay with this change, but would like a second opinion from @smoogipoo (see https://github.com/ppy/osu/pull/12376/commits/d0f30b7b422afe385fd1fc0de7a28684bf0ab7ef for reasoning).

smoogipoo replied:
Though it's super dodgy, I think I'm fine with this. I'd like this to not be a thing in the long term though.

peppy committed:
Reorganise and reword comments to make time override behaviour a bit clearer

peppy committed:
Merge conditionals for readability

peppy reviewed (approved):
I'm really happy with this direction. It reads a lot better and reduces the bar of entry to understanding what is going on with this logic.

The testing changes look correct and match my expectations better than what we had previously.

👍🏻

smoogipoo committed:
Merge branch 'master' into refactor-framed-replay-input-hander
